File: src/types/eda.ts
import type * as Plotly from "plotly.js";

export type ChartType =
  | "histogram"
  | "scatter"
  | "box"
  | "bar"
  | "line"
  | "violin"
  | "heatmap"
  | "density"
  | "pie"
  | "area"
  | "qq_plot";
export type LuminaDtype = "numeric" | "categorical" | "datetime" | "text" | "boolean";

export interface ChartConfig {
  chartId: string;
  chartType: ChartType;
  x: string | null;
  y: string | null;
  color: string | null;
  facet: string | null;
  aggregation?: string | null;
  values?: string | null;
  nbins?: number;
}

export interface ChartRequest {
  chart_type: ChartType;
  x?: string | null;
  y?: string | null;
  color?: string | null;
  facet?: string | null;
  aggregation?: string | null;
  values?: string | null;
  nbins?: number;
}

export interface ChartResponse {
  chart_id: string;
  chart_type: ChartType;
  plotly_figure: {
    data: Plotly.Data[];
    layout: Partial<Plotly.Layout>;
  };
  row_count: number;
  webgl: boolean;
  warnings: string[];
  downsampled: boolean;
  displayed_row_count: number | null;
}

export interface ShelfAssignment {
  shelf: "x" | "y" | "color" | "facet";
  columnName: string;
}

export interface DistributionRequest {
  column: string;
  group_by?: string | null;
  n_points?: number;
}

export interface KDETrace {
  group: string;
  x: number[];
  y: number[];
}

export interface DistributionResponse {
  column: string;
  group_by: string | null;
  traces: KDETrace[];
}

File: backend/app/models/eda.py
"""Pydantic request/response models for EDA chart endpoints."""

from pydantic import BaseModel, Field


class ChartRequest(BaseModel):
    """Chart generation request payload."""

    chart_type: str
    x: str | None = None
    y: str | None = None
    color: str | None = None
    facet: str | None = None
    nbins: int | None = None
    aggregation: str | None = None
    values: str | None = None


class ChartResponse(BaseModel):
    """Chart generation response payload."""

    chart_id: str
    chart_type: str
    plotly_figure: dict
    row_count: int
    webgl: bool
    warnings: list[str] = Field(default_factory=list)
    downsampled: bool = False
    displayed_row_count: int | None = None


class DownsampleRequest(BaseModel):
    """Request payload for chart-series downsampling."""

    x_column: str
    y_column: str
    threshold: int = Field(default=5000, ge=3)


class DownsampleResponse(BaseModel):
    """Response payload containing downsampled chart-series points."""

    x: list
    y: list
    original_count: int
    downsampled_count: int


class DistributionRequest(BaseModel):
    """Request payload for distribution comparison traces."""

    column: str
    group_by: str | None = None
    n_points: int = Field(default=200, ge=2)


class KDETrace(BaseModel):
    """A single density trace for one group."""

    group: str
    x: list[float]
    y: list[float]


class DistributionResponse(BaseModel):
    """Response payload for grouped density comparisons."""

    column: str
    group_by: str | None = None
    traces: list[KDETrace]
